### DISCUSSION TITLE
Can we change DataFetcher's execution flow?

### DISCUSSION THREAD

**Discussion by gabbigum** [graphql-java/graphql-java#2543] [Q&A]
Hello, everybody. I am curious if we **CAN** change the **DataFetcher**'s execution flow and if we can, **HOW**? 

What I am asking for is, can we first execute the **child** datafetcher (e.g. **reviews**) for this schema, as opposed to **default** parent to child datafetching?
```
type Query {
  shows: [Show]
}

type Show {
  showId: ID
  title: String
  reviews(minimumStarRating: Int): [Review]
}

type Review {
  starRating: Int
}
```

A visualisation might help more
```
    shows -  THEN resolve this field
    /
reviews(minimumStarRating: 3) - FIRST resolve this field through the datafetcher
```

What problems will this help us solve:
- We'll achieve **INNER-JOIN** like functionality in our queries, meaning that we'll be able to perform **correct** filtration, thus getting all the **shows** that have **reviews** with **minimumStarRating** of 3.

What we are **NOT** allowed:
- We cannot modify the schema
- Pre-loading is not an option
- We do not have direct access to DB

I know that this goes against the GraphQL spec for query execution, but I want to know if this can be achieved through GraphQL-java. Any recommendations or hints would be appreciated!

**Comment by dugenkui03** [comment]
I think it is not possible. Engine must complete parent field and pass parent field value to subField DataFetcher as `source`.

`Instrumentation` is a powerful tool, I'd like to suggest you filter `Query.shows` by `Instrumentation.beginFieldListComplete`.

**Comment by kvfedorov** [comment]
Again that is NOT correct:

>we'll be able to perform correct filtration, thus getting all the shows that have reviews with minimumStarRating of 3

Query you wrote reads as: return ALL shows, inside EACH show return only reviews with minimal rating 3. Anyone with general GraphQL knowledge would treat your schema that way.

Even if you would do what you want, it will NOT be spec complaint and thus will be confusing for clients of your service who does not know details of your implementation.

**Reply by asavov** [reply]
OKey, then the real question is: what's the correct graphql query (client site) and DataFetchers impl (backend) to achieve the following scenario?

High-level requirement:
- UI table with two columns: **Show** and **Reviews** (as defined by original comment):

== ALL Data ==
| Show    | Reviews |
 Apple      [review_A with rating 3*, review_B with rating 5*] 
 Orange   [review_O with rating 2*]
 Lemon    [review_L with rating 5*]

- Both columns are **Filterable**.

When Filter is `Review.Rating == 5*` then only **Apple, Lemon** Shows should be displayed (cause they match the filter)

== Filtered Data ==
| Show    | Reviews |
 Apple      [review_B with rating 5*] 
 Lemon    [review_L with rating 5*]

Challenges we want to find answer to and solve:
- GraphQL provides **OUTER-JOIN**-like functionality, as you mention **all** Shows are **always** presented **regardless** of the filtering on review rating.
- We need to provide **LEFT-INNER-JOIN**-like functionality: return only those Shows with Rating matching the specified filter.

What we explore as possible solution:
- A mechanism which excludes Shows from Parent result set when Child Review result set is null due to the filtering (as Orange was excluded).

What's the GraphQL way to achieve this?

Thanks in advance!

We are open to live chat discussion.

**Reply by kvfedorov** [reply]
As already suggested in previous discussion, just add _another_ query which would allow filtering:

```
input ShowsFilter {
    title: String
    rating: Int
    ... more filtering options as needed
}

type Query {
    showsFiltered(input: ShowsFilter): [Show]
    shows: [Show]
}
```
In `showsFiltered` resolver implementation you are free to do joins any way you need based on provided input

**Reply by gabbigum** [reply]
Yeah, I understand your point @kvfedorov, but doing this means that all the joins will be happening in the **parent** datafetcher. This approach totally eliminates all extensibility as the parent will be responsible for resolving fields which child datafetchers are supposed to resolve. Implementing it this way will make the child datafetchers totally useless, which IMO will lead to issues with using dataloaders'.

Thanks, we are still open for live discussion.

**Reply by kvfedorov** [reply]
I think you are mixing 2 different things here: filtering root entities based on some auxiliary information and retrieving nested data. The fact that you filter by rating of review does not imply that you will load any reviews to send to the client.

Consider this totally legit GraphQL schema:
```
type Show {
    title: String
    reviews(minRating: Int): [Review]
}

type Review {
    rating: Int
    show: Show
}

```
How would you implement your filtering here if it is all one query?

```
query {
    shows {
        reviews(minRating:3) {
            show {
                 reviews(minRating:5) {
                 ...
                 }
             }
         }
    }
}
```
From another angle: how in your setup would you write a query which loads only titles of shows of certain rating, but does not load any reviews along with them?

I'm not very interested in live discussion as I'm only doing this to offload it from actual project maintainers.

**Reply by lqc** [reply]
@gabbigum The parent is not responsible for resolving child data fetchers at all, but rather the field's arguments - which is the filtering criteria in this case. The arguments "stack" downwards, but I would never expect an argument to an inner field to affect the parent. 

I'm not also sure how you can do batching efficiently when you don't know show ids before you fetch all the reviews.

**Comment by gabbigum** [comment]
Thanks for all the feedback everyone. We'll rethink our schema and move forward.

**Comment by PeterPNBF** [comment]
I meet same problem, field4 in Parent depend on field2  and field3, or field4 and field3 depend on field2. 
```
type Parent{
  id: ID
  field1: String
  field2: Child1
  field3: Child2
  field4: Child3
}

type Child1{
    field6: String
}
type Child2{
    field7: String
}
type Child3{
    field8: String
}
```

**Reply by gabbigum** [reply]
Hi, what exactly the problem is? I find it hard to understand what you are trying to achieve, can you clarify it more?

**Comment by bbakerman** [comment]
> Query you wrote reads as: return ALL shows, inside EACH show return only reviews with minimal rating 3. Anyone with general GraphQL knowledge would treat your schema that way.

> Even if you would do what you want, it will NOT be spec complaint and thus will be confusing for clients of your service who does not know details of your implementation.


Now I am not sure this is technically true - its is true that generally graphql tends to act as an OUTER JOIN like system

But I am not sure there is a specification that says you MUST only do OUTER JOIN behavior - as far as I know the spec says nothing on the semantics on how data fetchers should act.

I will agree that  it MAY "be confusing for clients of your service who does not know details of your implementation." but I am pretty sure its not breaking an specification rules.

If you decide to look ahead and filter a parent field based on the arguments of a child field then it is not common.  But it could be done

It going to get confusing when you had say 4 types of filtering on child fields.  How do you combine that filtering, AND / OR ??

**Reply by kvfedorov** [reply]
Ok. I guess I'll admit did not check spec for that and it is just my mental model here after reading it few years back and using since then.

But even if spec permits other treatment, I still do not see how such treatment can be applied consistently to design even simplest examples here. So, either we have approach which permits to design any relation/filtering or run into query design issues right away. For me choice is obvious.
